dbms/src/Storages/StorageS3.h

```diff
@@ -0,0 +1,71 @@
+#pragma once
+
+#include <Storages/IStorage.h>
+#include <Poco/URI.h>
+#include <common/logger_useful.h>
+#include <ext/shared_ptr_helper.h>
+
+namespace DB
+{
+/**
+ * This class represents table engine for external S3 urls.
+ * It sends HTTP GET to server when select is called and
+ * HTTP PUT when insert is called.
+ */
+class StorageS3 : public ext::shared_ptr_helper<StorageS3>, public IStorage
+{
+public:
+    StorageS3(const Poco::URI & uri_,
+        const std::string & database_name_,
+        const std::string & table_name_,
+        const String & format_name_,
+        const ColumnsDescription & columns_,
```

Comment:
Maybe you also need constraints. Because all table engines support constraints.

Reply:
c03857b2aee08fee4616cecae6caf4dbd94555bf